src/app/users/get-user.ts

```diff
@@ -23,43 +22,40 @@ export const getUserForLogin = async ({
     { [SemanticAttributes.CODE_FUNCTION]: "getUserForLogin" },
     async () => {
       const user = await users.findById(userId as UserId)
-      if (user instanceof Error) {
-        return user
-      }
+      if (user instanceof Error) return user
       addAttributesToCurrentSpan({
         [ENDUSER_ALIAS]: user.username,
       })
       user.lastConnection = new Date()
 
-      // IP tracking logic could be extracted into domain
-      const ipConfig = getIpConfig()
-      if (ip && ipConfig.ipRecordingEnabled) {
-        const lastIP: IPType | undefined = user.lastIPs.find(
-          (ipObject: IPType) => ipObject.ip === ip,
-        )
-        if (lastIP) {
-          lastIP.lastConnection = user.lastConnection
-        } else if (ipConfig.proxyCheckingEnabled) {
-          const ipFetcher = IpFetcher()
-          const ipInfo = await ipFetcher.fetchIPInfo(ip as IpAddress)
-          if (!(ipInfo instanceof Error)) {
-            user.lastIPs.push({
-              ip,
-              ...ipInfo,
-              Type: ipInfo.type,
-              firstConnection: user.lastConnection,
-              lastConnection: user.lastConnection,
-            })
+      const updateIpInfo = async () => {
+        // IP tracking logic could be extracted into domain
+        const ipConfig = getIpConfig()
+        if (ip && ipConfig.ipRecordingEnabled) {
+          const lastIP: IPType | undefined = user.lastIPs.find(
+            (ipObject: IPType) => ipObject.ip === ip,
+          )
+          if (lastIP) {
+            lastIP.lastConnection = user.lastConnection
+          } else if (ipConfig.proxyCheckingEnabled) {
+            const ipFetcher = IpFetcher()
+            const ipInfo = await ipFetcher.fetchIPInfo(ip as IpAddress)
+            if (!(ipInfo instanceof Error)) {
+              user.lastIPs.push({
+                ip,
+                ...ipInfo,
+                Type: ipInfo.type,
+                firstConnection: user.lastConnection,
+                lastConnection: user.lastConnection,
+              })
+            }
           }
         }
+        await users.update(user)
```

Comment:
this is not concurrency safe. 

if the user get updated in the meantime (because `fetchIPInfo` is slow for instance, which is frequently the case), then we would overwrite the change.

cc @bodymindarts

Reply:
If I remember corretly we had the same discussion with addresses. The solution was to create its own repo https://github.com/GaloyMoney/galoy/blob/main/src/services/mongoose/wallet-on-chain-addresses.ts

Reply:
Exactly need to segregate the update paths either in the database level or logically via a repository. 

For the address situation we had to do the update database side because of the atomicity.

Reply:
Thinking more about it - other than segregating on DB level - this is probably the best intermediate solution.

Have a `UserConnectionsRepository` with a `persistNew` function that is atomic like the on-chain-addresses example.

We should remove `lastIPs` and `lastConnection` from the [user type](https://github.com/GaloyMoney/galoy/blob/main/src/domain/users/index.types.d.ts#L32) and remove those fields from the [`UsersRepository.update` function](https://github.com/GaloyMoney/galoy/blob/main/src/services/mongoose/users.ts#L60) so that there is no risk of concurrent update.

Like that we can continue to postpone the refactoring of the db schema itself until later. Thats the approach we've been taking so far.